src/test/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcherTest.java

```diff
@@ -79,4 +92,349 @@ public void testShouldUpdate() throws IOException {
+        FileRepositoryBuilder.create(project1Dir.resolve(".git").toFile())) {
+      repository.create();
+
+      createRepoTestFile(repository, firstProject, "testFile1");
+
+
+      fetcher = spy(new TestGitHubDataFetcher(github));
+
+      fetcher.addRepositoryInfoForTesting(firstProject, project1Dir);
+
+      localRepositoryFor(firstProject, 1);
+      assertEquals(fetcher.github(), github);
+    }
+
+    // Create the second project with second GitHub mock instance.
+    GitHubProject secondProject = new GitHubProject("test", "project2");
+    Path project2Dir = projectDir(secondProject.name());
+    GitHub github2 = mock(GitHub.class);
+    try (Repository repository =
+        FileRepositoryBuilder.create(project2Dir.resolve(".git").toFile())) {
+      repository.create();
+
+      createRepoTestFile(repository, secondProject, "testFile2");
+
+      fetcher = spy(new TestGitHubDataFetcher(github2));
+
+      fetcher.addRepositoryInfoForTesting(secondProject, project2Dir);
+
+      localRepositoryFor(secondProject, 2);
+      assertEquals(fetcher.github(), github2);
+    }
+
+    // Create the third project with second GitHub mock instance.
+    GitHubProject thirdProject = new GitHubProject("test", "project3");
+
+    Path project3Dir = projectDir(thirdProject.name());
+    try (Repository repository =
+        FileRepositoryBuilder.create(project3Dir.resolve(".git").toFile())) {
+      repository.create();
+      createRepoTestFile(repository, thirdProject, "testFile3");
+
+      fetcher.addRepositoryInfoForTesting(thirdProject, project3Dir);
+
+      localRepositoryFor(thirdProject, 3);
+      assertEquals(fetcher.github(), github2);
+    }
+  }
+
+  @Test
+  public void testMultiThreadReadAndUpdate() throws IOException, InterruptedException {
+    ScheduledExecutorService executorService = Executors.newScheduledThreadPool(2);
+
+    GitHubProject firstProject = new GitHubProject("test", "project");
+    Path project1Dir = projectDir(firstProject.name());
+    try (Repository repository =
+        FileRepositoryBuilder.create(project1Dir.resolve(".git").toFile())) {
+      repository.create();
+
+      createRepoTestFile(repository, firstProject, "testFile");
+
+      fetcher.addRepositoryInfoForTesting(firstProject, project1Dir);
+
+      CountDownLatch latch = new CountDownLatch(1);
+
+      Runnable cleanTask = () -> {
+        try {
+          fetcher.cleanup((url, repo, total) -> {
+            latch.countDown();
+            return true;
+          });
+          checkCleanUp(firstProject, 0);
+        } catch (IOException e) {
+          throw new IllegalStateException("Clean up exception");
+        }
+      };
+
+      GitHubProject secondProject = new GitHubProject("test", "project2");
+
+      Runnable loadRepoTask = () -> {
+        try {
+          assertTrue(latch.await(10, TimeUnit.SECONDS));
+          fetcher.localRepositoryFor(secondProject);
+          checkLocalRepository(secondProject, 1);
```

Comment:
By using `await()` the test tries to run the cleanup thread before this thread. Why does it need it? What if we remove the `await()` call? I also think that the `checkLocalRepository()` call should be moved to the parent thread.

Reply:
The scenario :
- First thread is running the `cleanup` task.
- Second thread is running the `loadLocalRepsotories` task.
We wanted to pause the first thread for some time. To confirm that the second thread will not cross the synchronized block until the first thread is complete. Hence, this approach. 

-`You are right `checkLocalRepository()` should be checked in the main thread. Have done this change

Reply:
> We wanted to pause the first thread for some time. To confirm that the second thread will not cross the synchronized block until the first thread is complete. Hence, this approach.

I am not sure why we need it. The most interesting part here is to let the threads run as they run.